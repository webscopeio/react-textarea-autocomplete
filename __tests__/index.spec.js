import React, { Component } from 'react';
import { shallow, mount } from 'enzyme';
import ReactTextareaAutocomplete from '../src';

// eslint-disable-next-line
const SmileItemComponent = ({ entity: { label, text } }) => (
  <div> {label} </div>
);

const Loading = () => <div>Loading...</div>;

describe('object-based items', () => {
  const mockedChangeFn = jest.fn();
  const mockedSelectFn = jest.fn();
  const mockedCaretPositionChangeFn = jest.fn();

  const rtaComponent = (
    <ReactTextareaAutocomplete
      listStyle={{ background: 'pink' }}
      itemStyle={{ background: 'green' }}
      containerStyle={{ background: 'orange' }}
      loaderStyle={{ background: 'blue' }}
      className="my-rta"
      containerClassName="my-rta-container"
      listClassName="my-rta-list"
      itemClassName="my-rta-item"
      loaderClassName="my-rta-loader"
      placeholder={'Write a message.'}
      value={'Controlled text'}
      onChange={mockedChangeFn}
      onSelect={mockedSelectFn}
      onCaretPositionChange={mockedCaretPositionChangeFn}
      style={{ background: 'red' }}
      loadingComponent={Loading}
      trigger={{
        ':': {
          output: item => `___${item.text}___`,
          dataProvider: () => [
            { id: 1, label: ':)', text: 'happy_face' },
            { id: 2, label: ':(', text: 'sad_face' },
          ],
          component: SmileItemComponent,
        },
      }}
    />
  );
  const rta = mount(rtaComponent);

  it('match the snapshot', () => {
    expect(shallow(rtaComponent)).toMatchSnapshot();
  });

  it('Textarea exists', () => {
    expect(rta.find('textarea')).toHaveLength(1);
  });

  it('After the trigger was typed, it should appear list of options', () => {
    rta
      .find('textarea')
      .simulate('change', { target: { value: 'some test :a' } });
    expect(rta.find('.rta__autocomplete')).toHaveLength(1);
  });

  it('match the snapshot of dropdown, list, and item', () => {
    expect(rta.find('.rta__autocomplete').getNode()).toMatchSnapshot();
    expect(
      rta.find('.rta__autocomplete .rta__list').getNode()
    ).toMatchSnapshot();
    expect(
      rta
        .find('.rta__autocomplete .rta__list .rta__item')
        .first()
        .getNode()
    ).toMatchSnapshot();
  });

  it('should display all items', () => {
    expect(rta.find('.rta__autocomplete .rta__list .rta__item')).toHaveLength(
      2
    );
  });

  it('items should be rendered within the component', () => {
    expect(
      rta
        .find('.rta__item')
        .first()
        .containsMatchingElement(<SmileItemComponent />)
    ).toEqual(true);
  });

  it('should invoke onChange handler', () => {
    expect(mockedChangeFn).toHaveBeenCalled();
  });

  it('should handle selection in textarea correctly', () => {
    rta.find('textarea').simulate('select');
    expect(mockedCaretPositionChangeFn).toHaveBeenCalled();
  });

  it('should close the autocomplete after mouse click', () => {
    const item = rta.find('.rta__entity').first();
    item.simulate('click');
    expect(rta.find('.rta__entity')).toHaveLength(0);
  });

  it('should invoke onChange handler after selection', () => {
    expect(mockedChangeFn).toHaveBeenCalledTimes(2);
  });

  it('should invoke onCaretPositionChange handler after selection', () => {
    expect(mockedCaretPositionChangeFn).toHaveBeenCalledTimes(2);
  });

  it('text in textarea should be changed', () => {
    expect(rta.find('textarea').node.value).toBe(
      '___happy_face___some test :a'
    );
  });
});

describe('string-based items w/o output fn', () => {
  const mockedChangeFn = jest.fn();
  const mockedCaretPositionChangeFn = jest.fn();

  const rtaComponent = (
    <ReactTextareaAutocomplete
      placeholder={'Write a message.'}
      value={'Controlled text'}
      onChange={mockedChangeFn}
      onCaretPositionChange={mockedCaretPositionChangeFn}
      className={'ownClassName'}
      style={{ background: 'red' }}
      loadingComponent={Loading}
      trigger={{
        ':': {
          dataProvider: () => Promise.resolve(['happy_face', 'sad_face']),
          component: SmileItemComponent,
        },
      }}
    />
  );
  const rta = mount(rtaComponent);

  it('match match the snapshot', () => {
    expect(shallow(rtaComponent)).toMatchSnapshot();
  });

  it('Textarea exists', () => {
    expect(rta.find('textarea')).toHaveLength(1);
  });

  it('After the trigger was typed, it should appear list of options', () => {
    rta
      .find('textarea')
      .simulate('change', { target: { value: 'some test :a' } });
    expect(rta.find('.rta__autocomplete')).toHaveLength(1);
  });

  it('should display all items', () => {
    expect(rta.find('.rta__autocomplete .rta__list .rta__item')).toHaveLength(
      2
    );
  });

  it('items should be rendered within the component', () => {
    expect(
      rta
        .find('.rta__item')
        .first()
        .containsMatchingElement(<SmileItemComponent />)
    ).toEqual(true);
  });

  it('should invoke onChange handler', () => {
    expect(mockedChangeFn).toHaveBeenCalled();
  });

  it('should handle selection in textarea correctly', () => {
    rta.find('textarea').simulate('select');
    expect(mockedCaretPositionChangeFn).toHaveBeenCalled();
  });

  it('should close the autocomplete after mouse click', () => {
    const item = rta.find('.rta__entity').first();
    item.simulate('click');
    expect(rta.find('.rta__entity')).toHaveLength(0);
  });

  it('should invoke onChange handler after selection', () => {
    expect(mockedChangeFn).toHaveBeenCalledTimes(2);
  });

  it('should invoke onCaretPositionChange handler after selection', () => {
    expect(mockedCaretPositionChangeFn).toHaveBeenCalledTimes(2);
  });

  it('text in textarea should be changed', () => {
    expect(rta.find('textarea').node.value).toBe(':happy_face:some test :a');
  });
});

describe('string-based items with output fn', () => {
  const mockedChangeFn = jest.fn();
  const mockedCaretPositionChangeFn = jest.fn();
  const rtaComponent = (
    <ReactTextareaAutocomplete
      placeholder={'Write a message.'}
      value={'Controlled text'}
      onChange={mockedChangeFn}
      onCaretPositionChange={mockedCaretPositionChangeFn}
      className={'ownClassName'}
      style={{ background: 'red' }}
      loadingComponent={Loading}
      trigger={{
        ':': {
          output: item => `__${item}__`,
          dataProvider: () => Promise.resolve(['happy_face', 'sad_face']),
          component: SmileItemComponent,
        },
      }}
    />
  );
  const rta = mount(rtaComponent);

  it('match match the snapshot', () => {
    expect(shallow(rtaComponent)).toMatchSnapshot();
  });

  it('Textarea exists', () => {
    expect(rta.find('textarea')).toHaveLength(1);
  });

  it('After the trigger was typed, it should appear list of options', () => {
    rta
      .find('textarea')
      .simulate('change', { target: { value: 'some test :a' } });
    expect(rta.find('.rta__autocomplete')).toHaveLength(1);
  });

  it('should display all items', () => {
    expect(rta.find('.rta__autocomplete .rta__list .rta__item')).toHaveLength(
      2
    );
  });

  it('items should be rendered within the component', () => {
    expect(
      rta
        .find('.rta__item')
        .first()
        .containsMatchingElement(<SmileItemComponent />)
    ).toEqual(true);
  });

  it('should invoke onChange handler', () => {
    expect(mockedChangeFn).toHaveBeenCalled();
  });

  it('should handle selection in textarea correctly', () => {
    rta.find('textarea').simulate('select');
    expect(mockedCaretPositionChangeFn).toHaveBeenCalled();
  });

  it('should close the autocomplete after mouse click', () => {
    const item = rta.find('.rta__entity').first();
    item.simulate('click');
    expect(rta.find('.rta__entity')).toHaveLength(0);
  });

  it('should invoke onChange handler after selection', () => {
    expect(mockedChangeFn).toHaveBeenCalledTimes(2);
  });

  it('should invoke onCaretPositionChange handler after selection', () => {
    expect(mockedCaretPositionChangeFn).toHaveBeenCalledTimes(2);
  });

  it('text in textarea should be changed', () => {
    expect(rta.find('textarea').node.value).toBe('__happy_face__some test :a');
  });
});

describe('using ref to the ReactTextareaAutocomplete to call methods', () => {
  const mockedChangeFn = jest.fn();

  const rtaComponent = (
    <ReactTextareaAutocomplete
      placeholder={'Write a message.'}
      value={'Controlled text'}
      onChange={mockedChangeFn}
      className={'ownClassName'}
      style={{ background: 'red' }}
      loadingComponent={Loading}
      trigger={{
        ':': {
          dataProvider: () => Promise.resolve(['happy_face', 'sad_face']),
          component: SmileItemComponent,
        },
      }}
    />
  );

  const rtaWrapper = mount(rtaComponent);
  const rtaWrapperRef = rtaWrapper.instance();

  it('should get the correct caret position initially', () => {
    const actual = rtaWrapperRef.getCaretPosition();
    const expected = 0;

    expect(actual).toBe(expected);
  });

  it('should set the caret position correctly', () => {
    const CARET_POSITION_TO_SET = 5;

    rtaWrapperRef.setCaretPosition(CARET_POSITION_TO_SET);

    const actual = rtaWrapperRef.getCaretPosition();
    const expected = CARET_POSITION_TO_SET;

    expect(actual).toBe(expected);
  });
});
