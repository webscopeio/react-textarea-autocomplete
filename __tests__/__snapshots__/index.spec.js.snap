// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`object-based items match the snapshot 1`] = `
<div
  className="rta  my-rta-container"
  style={
    Object {
      "background": "orange",
    }
  }
>
  <textarea
    className="rta__textarea my-rta"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onScroll={[Function]}
    onSelect={[Function]}
    placeholder="Write a message."
    style={
      Object {
        "background": "red",
      }
    }
    value="Controlled text"
  />
</div>
`;

exports[`object-based items match the snapshot of dropdown, list, and item 1`] = `
<div
  class="rta__autocomplete "
  style="top: 0px;"
>
  <ul
    class="rta__list my-rta-list"
  >
    <li
      class="rta__item my-rta-item"
      style="background: green;"
    >
      <div
        class="rta__entity rta__entity--selected"
        role="button"
        tabindex="0"
      >
        <div>
           
          :)
           
        </div>
      </div>
    </li>
    <li
      class="rta__item my-rta-item"
      style="background: green;"
    >
      <div
        class="rta__entity "
        role="button"
        tabindex="0"
      >
        <div>
           
          :(
           
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`object-based items match the snapshot of dropdown, list, and item 2`] = `
<ul
  class="rta__list my-rta-list"
>
  <li
    class="rta__item my-rta-item"
    style="background: green;"
  >
    <div
      class="rta__entity rta__entity--selected"
      role="button"
      tabindex="0"
    >
      <div>
         
        :)
         
      </div>
    </div>
  </li>
  <li
    class="rta__item my-rta-item"
    style="background: green;"
  >
    <div
      class="rta__entity "
      role="button"
      tabindex="0"
    >
      <div>
         
        :(
         
      </div>
    </div>
  </li>
</ul>
`;

exports[`object-based items match the snapshot of dropdown, list, and item 3`] = `
<li
  class="rta__item my-rta-item"
  style="background: green;"
>
  <div
    class="rta__entity rta__entity--selected"
    role="button"
    tabindex="0"
  >
    <div>
       
      :)
       
    </div>
  </div>
</li>
`;

exports[`object-based items with keys match the snapshot 1`] = `
<div
  className="rta  my-rta-container"
  style={
    Object {
      "background": "orange",
    }
  }
>
  <textarea
    className="rta__textarea my-rta"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onScroll={[Function]}
    onSelect={[Function]}
    placeholder="Write a message."
    style={
      Object {
        "background": "red",
      }
    }
    value="Controlled text"
  />
</div>
`;

exports[`object-based items with keys match the snapshot of dropdown, list, and item 1`] = `
<div
  class="rta__autocomplete "
  style="top: 0px;"
>
  <ul
    class="rta__list my-rta-list"
  >
    <li
      class="rta__item my-rta-item"
      style="background: green;"
    >
      <div
        class="rta__entity rta__entity--selected"
        role="button"
        tabindex="0"
      >
        <div>
           
          :)
           
        </div>
      </div>
    </li>
    <li
      class="rta__item my-rta-item"
      style="background: green;"
    >
      <div
        class="rta__entity "
        role="button"
        tabindex="0"
      >
        <div>
           
          :(
           
        </div>
      </div>
    </li>
    <li
      class="rta__item my-rta-item"
      style="background: green;"
    >
      <div
        class="rta__entity "
        role="button"
        tabindex="0"
      >
        <div>
           
          :|
           
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`object-based items with keys match the snapshot of dropdown, list, and item 2`] = `
<ul
  class="rta__list my-rta-list"
>
  <li
    class="rta__item my-rta-item"
    style="background: green;"
  >
    <div
      class="rta__entity rta__entity--selected"
      role="button"
      tabindex="0"
    >
      <div>
         
        :)
         
      </div>
    </div>
  </li>
  <li
    class="rta__item my-rta-item"
    style="background: green;"
  >
    <div
      class="rta__entity "
      role="button"
      tabindex="0"
    >
      <div>
         
        :(
         
      </div>
    </div>
  </li>
  <li
    class="rta__item my-rta-item"
    style="background: green;"
  >
    <div
      class="rta__entity "
      role="button"
      tabindex="0"
    >
      <div>
         
        :|
         
      </div>
    </div>
  </li>
</ul>
`;

exports[`object-based items with keys match the snapshot of dropdown, list, and item 3`] = `
<li
  class="rta__item my-rta-item"
  style="background: green;"
>
  <div
    class="rta__entity rta__entity--selected"
    role="button"
    tabindex="0"
  >
    <div>
       
      :)
       
    </div>
  </div>
</li>
`;

exports[`object-based items without keys and custom unique generator match the snapshot 1`] = `
<div
  className="rta  "
>
  <textarea
    className="rta__textarea "
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onScroll={[Function]}
    onSelect={[Function]}
    placeholder="Write a message."
    value="Controlled text"
  />
</div>
`;

exports[`string-based items w/o output fn match match the snapshot 1`] = `
<div
  className="rta  "
>
  <textarea
    className="rta__textarea ownClassName"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onScroll={[Function]}
    onSelect={[Function]}
    placeholder="Write a message."
    style={
      Object {
        "background": "red",
      }
    }
    value="Controlled text"
  />
</div>
`;

exports[`string-based items with output fn match match the snapshot 1`] = `
<div
  className="rta  "
>
  <textarea
    className="rta__textarea ownClassName"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onScroll={[Function]}
    onSelect={[Function]}
    placeholder="Write a message."
    style={
      Object {
        "background": "red",
      }
    }
    value="Controlled text"
  />
</div>
`;
